File: sentimental-credit/credit.py
# TODO

import re


def main():
    # get card number from user - do as a while loop, to re prompt if input is invalid
    card_number = (input("Enter a card number: "))

    # amex check - checking if card is 15 in length and starts with 34 or 37
    if re.search("[0-9]{15}", card_number) and (re.search("^34", card_number) or re.search("^37", card_number)):
        # if it is and if checksum then true, print amex
        if checksum(card_number):
            print("AMEX")
        else:
            print("INVALID")
            exit()
    # visa check 1- check if card is 16 digits and starts with 4
    elif re.search("[0-9]{16}", card_number) and re.search("^4", card_number):
        # if it is and if checksum then true, print Visa
        if checksum(card_number):
            print("VISA")
        else:
            print("INVALID")
            exit()

    # Visa Check 2 - checking if card is 13 digits and starts with 4
    elif re.search("[0-9]{13}", card_number) and re.search("^4", card_number):
        # if it is and if checksum then true, print Visa
        if checksum(card_number):
            print("VISA")
        else:
            print("INVALID")
            exit()
    # mastercard check - checking if 16 digit and start with 51, 52, 53 or 54
    elif re.search("[0-9]{16}", card_number) and (re.search("^51", card_number) or re.search("^52", card_number) or re.search("^53", card_number) or re.search("^54", card_number) or re.search("^55", card_number)):
        # if it is and if checksum then true, print Mastercard
        if checksum(card_number):
            print("MASTERCARD")
        else:
            print("INVALID")
            exit()

    else:
        print("INVALID")
        exit()


def checksum(number):

    # convert cardnumber into an array of the digits- map iterates over the string converting, adding a function of my choice (in this case int)
    # then list() converts the map object to a list
    card_list = list((map(int, number)))

    # step 1 of luhn's algorythm
    luhns_one = []
    num = len(number)

    # if the card length number is odd, do the following:
    if num % 2 != 0:
        # iterate over card_list backwards, every other number
        for i in range(num-1, 0, -2):

            # times i-2 by 2 as part of the algorithm
            num2 = card_list[i-1]
            num2 *= 2

            # if i-2 is just one digit, append that to luhns_one
            if (num2 / 10) < 1:
                luhns_one.append(num2)
            # if i-2 is two digits, append the digits seperately to the array, so we can add them all up
            else:
                luhns_one.append(1)
                luhns_one.append(num2 % 10)

        # now add up all the numbers in the luhns_one list by looping over it
        luhns_one_total = 0

        for i in luhns_one:
            luhns_one_total += i

        # part 2 of luhns algorithm -  add the sum of the numbers that we didn't multiple by 2 + add my luhns_one_value to that
        luhns_two_total = luhns_one_total
        for i in range(num, 0, -2):
            luhns_two_total += card_list[i-1]

        # finally, im checking if the total number ends in zero - if it does, we have a valid number and we return true
        if luhns_two_total % 10 == 0:
            return True
        else:
            return False

    # if the card number length is even, do the following:
    elif num % 2 == 0:
        # iterate over card_list backwards, every other number
        for i in range(num, 0, -2):

            # times i-2 by 2 as part of the algorithm
            num2 = card_list[i-2]
            num2 *= 2

            # if i-2 is just one digit, append that to luhns_one
            if (num2 / 10) < 1:
                luhns_one.append(num2)
            # if i-2 is two digits, append the digits seperately to the array, so we can add them all up
            else:
                luhns_one.append(1)
                luhns_one.append(num2 % 10)

        # now add up all the numbers in the luhns_one list by looping over it
        luhns_one_total = 0

        for i in luhns_one:
            luhns_one_total += i

        # part 2 of luhns algorithm -  add the sum of the numbers that we didn't multiple by 2 + add my luhns_one_value to that
        luhns_two_total = luhns_one_total
        for i in range(num, 0, -2):
            luhns_two_total += card_list[i-1]

        # finally, im checking if the total number ends in zero - if it does, we have a valid number and we return true
        if luhns_two_total % 10 == 0:
            return True
        else:
            return False


if __name__ == '__main__':
    main()

File: credit/credit.c

#include <cs50.h>
#include <stdio.h>



//checksum function declaration- which will take the card number as an argument:
bool checksum_odd(long card);
bool checksum_even(long card);

int main(void)
{

    long card;
    do //get card number, where it must be more than 0
    {
        card = get_long("Please enter card number:");
    }
    while (card < 0);

    // amex check
    //checking if card is 15 in length and starts with 34 or 37
    if((card/100000000000000 >0 && card /1000000000000000 <1) && (card/10000000000000 == 34 || card/10000000000000 == 37))
    {
        //if so, put it through the checksum function and print amex if true
        bool check = checksum_odd(card);
        if (check == true)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    // Visa Check 16  - checking if card is 16 digits and starts with 4
    else if((card/1000000000000000 >0 && card /10000000000000000 <1) && (card/1000000000000000 == 4 ))
    {
       // putting through checksum
         bool check = checksum_even(card);
         if (check == true)
         {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }


        // Visa Check 13 - checking if card is 13 digits and starts with 4
    else if((card/1000000000000 >0 && card /10000000000000 <1) && (card/1000000000000 == 4 ))
    {
        //putting through checksum
        bool check = checksum_odd(card);
        if (check == true)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }

    // mastercard check - must be 16 digit and start with 51, 52, 53 or 54
    else if((card/1000000000000000 >0 && card /10000000000000000 <1) && (card/100000000000000 == 51 || card/100000000000000 == 52 || card/100000000000000 == 53 || card/100000000000000 == 54 || card/100000000000000 == 55 ))
    {
    //putting through checksum
        bool check = checksum_even(card);
         if (check == true)
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else{
        printf("INVALID\n");
    }


} //int main bracket



//checksum function for odd numbers

bool checksum_odd(long card)
{
    long x = card;
    bool test = false;

    //working out how long the card is, keep dividing by 10 until it equals 0, while adding 1 to x each round
    long length = 0;
    while(x!= 0)
    {
        x /=10;
        length +=1;
    }

    int z = length;
    int digitarray_length;
    int digitarray_length_forlooping;
    int addto_array;
    int luhns_check;
    int leftover = 0;
    int leftover_sum = 0;


    // if length of the card is an odd number
    if(!(z % 2 == 0))
    {
        // working out how long the array should be of every other number
        digitarray_length = ((z-1)/2);
        digitarray_length_forlooping = digitarray_length;
        int size = digitarray_length;
        int everyothernumber_array[size];


        //while length is greater than 0, modulo 10 the card - and add that value to leftover sum
        while(z>0)
        {
            leftover  = card % 10;
            leftover_sum  += leftover;
            card = card / 10;
            z-=1;
        // then catch the next number in an array of the every other numbers and repeat until we've gone through all the numbers
            if(z>0)
            {
                addto_array = card % 10;
                // its this line that doesnt work which must mean i dont know how to set arrays
                everyothernumber_array[size-1] = addto_array;
                card = card / 10;
                size -=1;
            }
            else {
                    z = 0;
                }
        }

        int product_digits = 0;
        // work out the sum of the digits in the array - this requires iterating over each array element, so we have a loop in a loop and we're adding the product digits together
        for(x=0; x < digitarray_length_forlooping; x++)
        {
            everyothernumber_array[x] *=2;
            while (everyothernumber_array[x] != 0)
            {
                product_digits += everyothernumber_array[x] %10;
                everyothernumber_array[x] /= 10;

            }
        }

        // here i'm doing luhns check, if it's 0 we know its a valid card number
        luhns_check = product_digits + leftover_sum;
        if(luhns_check % 10 == 0)
        {
            test = true;

        }
        else
        {
            test = false;
        }
    }
    return test;


}
    //checksumodd end

// here's the function for an even number  -sames steps, but changes made to accurately get every other number etc, i should move this into one big fucntion really
bool checksum_even(long card)
{
    long x = card;
    bool test = false;
    long length = 0;
    while(x!= 0)
    {
        x /=10;
        length +=1;
    }

    int z = length;
    int digitarray_length;
    int digitarray_length_forlooping;
    int addto_array;
    int luhns_check;
    int leftover = 0;
    int leftover_sum = 0;



    if(z % 2 == 0)
    {
        digitarray_length = (z/2);
        digitarray_length_forlooping = digitarray_length;
        int size = digitarray_length;
        int everyothernumber_array[size];


        while(z>0)
        {
            leftover  = card % 10;
            leftover_sum  += leftover;
            card = card / 10;
            z-=1;

            if(z>0)
            {
                addto_array = card % 10;
                // its this line that doesnt work which must mean i dont know how to set arrays
                everyothernumber_array[size-1] = addto_array;
                card = card / 10;
                size -=1;
            }
            else {
                    z = 0;
                }
        }

        int product_digits = 0;
// new section adding the products digits
        for(x=0; x < digitarray_length_forlooping; x++)
        {
            everyothernumber_array[x] *=2;
            while (everyothernumber_array[x] != 0)
            {
                product_digits += everyothernumber_array[x] %10;
                everyothernumber_array[x] /= 10;

            }
        }
        luhns_check = product_digits + leftover_sum;
        if(luhns_check % 10 == 0)
        {
            test = true;

        }
        else
        {
            test = false;
        }
    }
    return test;


}





